bpf/kmesh/ads/include/circuit_breaker.h

```diff
@@ -0,0 +1,172 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Kmesh */
+
+#include "bpf_log.h"
+#include "kmesh_common.h"
+#include "bpf_common.h"
+
+#ifndef __KMESH_CIRCUIT_BREAKER_H__
+#define __KMESH_CIRCUIT_BREAKER_H__
+
+#define CLUSTER_NAME_MAX_LEN BPF_DATA_MAX_LEN
+
+struct cluster_stats {
+    __u32 active_connections;
+};
+
+struct cluster_stats_key {
+    __u64 netns_cookie;
+    __u32 cluster_id;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(key_size, sizeof(struct cluster_stats_key));
+    __uint(value_size, sizeof(struct cluster_stats));
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+    __uint(max_entries, MAP_SIZE_OF_CLUSTER);
+} map_of_cluster_stats SEC(".maps");
+
+struct cluster_sock_data {
+    __u32 cluster_id;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_SK_STORAGE);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+    __type(key, int);
+    __type(value, struct cluster_sock_data);
+} map_of_cluster_sock SEC(".maps");
+
+static inline void update_cluster_active_connections(const struct cluster_stats_key *key, int delta)
+{
+    if (!key) {
+        return;
+    }
+    struct cluster_stats *stats = NULL;
+    stats = kmesh_map_lookup_elem(&map_of_cluster_stats, key);
+    if (!stats) {
+        struct cluster_stats init_value = {0};
+        bpf_map_update_elem(&map_of_cluster_stats, key, &init_value, BPF_NOEXIST);
+        stats = kmesh_map_lookup_elem(&map_of_cluster_stats, key);
+    }
+
+    if (!stats) {
+        BPF_LOG(ERR, KMESH, "failed to get cluster stats");
+        return;
+    }
+    if (delta < 0 && -delta > stats->active_connections) {
+        BPF_LOG(ERR, KMESH, "invalid delta update");
+        return;
+    }
+
+    __sync_fetch_and_add(&stats->active_connections, delta);
```

Comment:
It seems it only protect `stats->active_connections`, not the whole `stats`.  At the same time, if a nother prog fetches stats, it can also update the value?

Reply:
since it's atomic, update will be before-or-after, only one processor will update stats->active_connections at a time.

Reply:
Yes, map_lookup is a memory reference, so this atomic op is concurrent safe